pymc/gp/hsgp.py

```diff
@@ -0,0 +1,265 @@
+        return phi[:, i:], pt.sqrt(psd[i:])
+
+    @staticmethod
+    def _eigendecomposition(X, L, m, n_dims):
+        """Construct the eigenvalues and eigenfunctions of the Laplace operator."""
+        m_star = pt.prod(m)
+        S = np.meshgrid(*[np.arange(1, 1 + m[d]) for d in range(n_dims)])
+        S = np.vstack([s.flatten() for s in S]).T
+        eigvals = pt.square((np.pi * S) / (2 * L))
+        phi = pt.ones((X.shape[0], m_star))
+        for d in range(n_dims):
+            c = 1.0 / np.sqrt(L[d])
+            phi *= c * pt.sin(pt.sqrt(eigvals[:, d]) * (pt.tile(X[:, d][:, None], m_star) + L[d]))
+        omega = pt.sqrt(eigvals)
+        return omega, phi, m_star
+
+    def prior(self, name: str, X: Union[np.ndarray, pt.TensorVariable], *args, **kwargs):
+        R"""
+        Returns the (approximate) GP prior distribution evaluated over the input locations `X`.
+
+        Parameters
+        ----------
+        name: string
+            Name of the random variable
+        X: array-like
+            Function input values.
+        dims: None
+            Dimension name for the GP random variable.
+        """
+
+        phi, sqrt_psd = self.prior_components(X)
+
+        if self.parameterization == "noncentered":
+            self.beta = pm.Normal(f"{name}_hsgp_coeffs_", size=sqrt_psd.size)
+            f = self.mean_func(X) + phi @ (self.beta * sqrt_psd)
+
+        elif self.parameterization == "centered":
+            self.beta = pm.Normal(f"{name}_hsgp_coeffs_", sigma=sqrt_psd, size=sqrt_psd.size)
+            f = self.mean_func(X) + phi @ self.beta
+
+        self.f = pm.Deterministic(name, f, dims=kwargs.get("dims"))
+        return self.f
+
+    def _build_conditional(self, Xnew):
+        try:
+            beta = self.beta
+        except AttributeError:
+            raise ValueError(
+                "Prior is not set, can't create a conditional.  Call `.prior(name, X)` first."
+            )
+
+        Xnew, _ = self.cov_func._slice(Xnew)
+        omega, phi, _ = self._eigendecomposition(Xnew, self.L, self.m, self.n_dims)
+        i = int(self.drop_first == True)
+
+        if self.parameterization == "noncentered":
+            psd = self.cov_func.power_spectral_density(omega)
+            return self.mean_func(Xnew) + phi[:, i:] @ (beta * pt.sqrt(psd[i:]))
+
+        elif self.parameterization == "centered":
+            return self.mean_func(Xnew) + phi[:, i:] @ beta
+
+    def conditional(self, name: str, Xnew: Union[np.ndarray, pt.TensorVariable], *args, **kwargs):
+        R"""
+        Returns the (approximate) conditional distribution evaluated over new input locations
+        `Xnew`.  If using the
+
+        Parameters
+        ----------
+        name: string
+            Name of the random variable
+        Xnew: array-like
+            Function input values.
+        kwargs: dict-like
+            Optional arguments such as `dims`.
+        """
+        fnew = self._build_conditional(Xnew)
```

Comment:
Should it use the same self.prior_components function?

Reply:
Is this snippet correct?
```python
phi, sqrt_psd = gp.prior_components(Xnew)

if self.parameterization == "centered":
    return phi @ self.beta
else:
    return phi @ (self.beta * sqrt_psd])
```

Reply:
For prediction you only need to compute a new basis phi.  The power spectral density doesn't change, it just depends on the number of basis vectors used.  You use the posterior for beta.  Ive refactored it a bit to hopefully make this more clear.  It also works with `pm.set_data` now, like if it were a linear regression!  I renamed `prior_components` to `prior_linearized` and added an example to the docstring there.  let me know if its clearer now